app/code/community/Apruve/ApruvePayment/Model/PaymentMethod.php

```diff
@@ -24,161 +24,219 @@
+	 * @param float $amount
+	 * @return Mage_Payment_Model_Abstract|void
+	 * @throws Mage_Core_Exception
+	 */
+	public function authorize(Varien_Object $payment, $amount)
+	{
+		parent::authorize($payment, $amount);
+		Mage::helper('apruvepayment')->logException('Authorize...');
+
+		$additionalInformation = $payment->getAdditionalInformation();
+		$token = $additionalInformation['aprt'];
+
+		/** @var Mage_Sales_Model_Order $order */
+		$order = $payment->getOrder();
+
+		/** @var Apruve_ApruvePayment_Model_Api_Rest_Order $orderApi */
+		$orderApi = Mage::getModel('apruvepayment/api_rest_order');
+
+
+		$updateResult = $orderApi->updateOrder($token, $order);
+		if (!$updateResult || !$updateResult['success']) {
+			Mage::throwException('Couldn\'t update order in Apruve.');
+		}
+
+		$payment->setTransactionId($updateResult['response']['id']);
+		$payment->setIsTransactionClosed(0);
+
+		return $this;
+	}
+
+	/**
+	 * Captures a payment
+	 *
+	 * @param   Varien_Object $payment
+	 * @return  bool
+	 * @throws  Mage_Core_Exception
+	 */
+	public function capture(Varien_Object $payment, $amount)
+	{
+		parent::capture($payment, $amount);
+		Mage::helper('apruvepayment')->logException('Capture...');
+
+		if ($amount <= 0) {
+			Mage::throwException(Mage::helper('paygate')->__('Invalid amount for capture.'));
+		}
+
+		$payment->setSkipTransactionCreation(true);
+		return $this;
+	}
+
+
+	/**
+	 * Create a refund
+	 *
+	 * @param   Varien_Object $payment
+	 * @return  bool
+	 * @throws  Mage_Core_Exception
+	 */
+	public function refund(Varien_Object $payment, $amount){
+		Mage::helper('apruvepayment')->logException('Refund...');
+
+		// Get Magento Order
+		$order = $payment->getOrder();
+		$magentoOrderId = $order->getIncrementId();
+		$apruveEntity = Mage::getModel('apruvepayment/entity')->loadByOrderId($magentoOrderId, 'magento_id');
+		$apruveOrderId = $apruveEntity->getApruveId();
+
+		$magentoInvoices = [];
+
+		$invoiceCollection = $order->getInvoiceCollection();
+		foreach($invoiceCollection as $invoice):
+			$magentoInvoices[] =  $invoice->getIncrementId();
+		endforeach;
+
+		// Get API objects
+		$orderApi = Mage::getModel('apruvepayment/api_rest_order');
+		$invoiceApi = Mage::getModel('apruvepayment/api_rest_invoice');
+
+		// User Order to Get Apruve invoices
+		$apruveInvoices = $orderApi->getInvoices($apruveOrderId);
+
+		$validInvoice = null;
+		$totalAmountCents = 0;
+
+		foreach ($apruveInvoices as $invoice){
+			$totalAmountCents += $invoice['amount_cents'];
+			if (in_array($invoice['merchant_invoice_id'], $magentoInvoices)){
+				$validApruveInvoiceId = $invoice['id'];
+			}
+		}
+
+		Mage::helper('apruvepayment')->logException('$totalAmountCents: ' . $totalAmountCents);
+		Mage::helper('apruvepayment')->logException('$validInvoice: ' . $validInvoice);
+		if ($totalAmountCents >= ($amount * 100)  && !empty($validApruveInvoiceId)){
+			$result = $invoiceApi->refundInvoice($validApruveInvoiceId, $amount);
```

Comment:
The idea was to iterate through the Apruve invoices to find the invoice that matches the Magento invoice that is having a refund applied to it (based on the merchant_invoice_id). This is assuming that the $payment object gives you access to the invoice.